codex-rs/tui/src/resume_picker.rs

```diff
@@ -0,0 +1,453 @@
+            let anchor = self.backstack.get(self.page_index).cloned().flatten();
+            self.current_anchor = anchor.clone();
+            self.load_page(anchor.as_ref()).await?;
+        }
+        Ok(())
+    }
+
+    async fn next_page(&mut self) -> Result<()> {
+        if let Some(next) = self.next_cursor.clone() {
+            // Record the anchor for the page we are moving to at index new_index
+            let new_index = self.page_index + 1;
+            if self.backstack.len() <= new_index {
+                self.backstack.resize(new_index + 1, None);
+            }
+            self.backstack[new_index] = Some(next.clone());
+            self.current_anchor = Some(next.clone());
+            self.page_index = new_index;
+            let anchor = self.current_anchor.clone();
+            self.load_page(anchor.as_ref()).await?;
+        }
+        Ok(())
+    }
+
+    async fn load_page(&mut self, anchor: Option<&Cursor>) -> Result<()> {
+        let page = RolloutRecorder::list_conversations(&self.codex_home, 25, anchor).await?;
+        self.next_cursor = page.next_cursor.clone();
+        self.all_rows = to_rows(page);
+        self.apply_filter();
+        // reset selection on new page
+        self.selected = 0;
+        Ok(())
+    }
+
+    fn apply_filter(&mut self) {
+        if self.query.is_empty() {
+            self.filtered_rows = self.all_rows.clone();
+        } else {
+            let q = self.query.to_lowercase();
+            self.filtered_rows = self
+                .all_rows
+                .iter()
+                .filter(|r| r.preview.to_lowercase().contains(&q))
+                .cloned()
+                .collect();
+        }
+        if self.selected >= self.filtered_rows.len() {
+            self.selected = self.filtered_rows.len().saturating_sub(1);
+        }
+        self.request_frame();
+    }
+}
+
+fn to_rows(page: ConversationsPage) -> Vec<Row> {
+    use std::cmp::Reverse;
+    let mut rows: Vec<Row> = page
+        .items
+        .into_iter()
+        .filter_map(|it| head_to_row(&it))
+        .collect();
+    // Ensure newest-first ordering within the page by timestamp when available.
+    let epoch = Utc.timestamp_opt(0, 0).single().unwrap_or_else(Utc::now);
+    rows.sort_by_key(|r| Reverse(r.ts.unwrap_or(epoch)));
+    rows
+}
+
+fn head_to_row(item: &ConversationItem) -> Option<Row> {
+    let mut ts: Option<DateTime<Utc>> = None;
+    if let Some(first) = item.head.first()
+        && let Some(t) = first.get("timestamp").and_then(|v| v.as_str())
+        && let Ok(parsed) = chrono::DateTime::parse_from_rfc3339(t)
+    {
+        ts = Some(parsed.with_timezone(&Utc));
+    }
+
+    let preview = find_first_user_text(&item.head)?;
+    if preview.trim().is_empty() {
+        return None;
+    }
+    Some(Row {
+        path: item.path.clone(),
+        preview,
+        ts,
+    })
+}
+
+fn find_first_user_text(head: &[serde_json::Value]) -> Option<String> {
```

Comment:
Please add a docstring to this function: the code is a bit dense so it's helpful to have a comment to understand what the strategy is here.